h/h_api/bulk_api/command_batch.py

```diff
@@ -0,0 +1,88 @@
+"""Batching and dispatching jobs to work on."""
+
+from contextlib import contextmanager
+
+from h.h_api.exceptions import CommandSequenceException
+
+
+class CommandBatch:
+    """A batch of commands.
+
+    This object will group similar commands together and decide when they
+    should be executed. The number of commands should always be a number we
+    are comfortable dealing with in memory at once.
+    """
+
+    def __init__(self, on_flush, batch_size=100):
+        """
+        :param on_flush: Callback to call when it's time to execute
+        :param batch_size: The max size of a batch
+        """
+
+        self.on_flush = on_flush
+
+        self.batch = None
+        self.batch_size = batch_size
+        self.finished_tasks = set()
+        self.current_task = None
+
+    def flush(self):
+        """Flush the current batch of commands to the callback."""
+
+        if not self.batch:
+            self.batch = None
+            return
+
+        command_type, data_type = self.current_task
+        self.on_flush(command_type, data_type, self.batch)
+
+        self.batch = None
+
+    @contextmanager
+    def add(self, command):
+        """
+        A context manager for adding a command to the batch.
+
+        Making this a context manager allows us to do the prep-work we need
+        to before and after a command is added.
```

Comment:
I don't understand this part. Why is it a context manager? How will it actually be used in the end?

Reply:
It's a slightly finickety ordering problem to do with id references. If we have a bunch of groups with user provided references, and we add a group_membership which uses that reference we need to flush and process the groups before we can fill out the group_membership reference to the group.

We can't do this before, as groups aren't processed yet. We can't do it after, because the batch would contain an unprocessable command (which could get flushed). The only option is to do it during so we:

1. Flush old stuff and populate references
2. Populate the new item, potentially based on the flushed items
3. Add the new item to a fresh batch

There's no way you'd get this from this PR though. I did start with a plain add, but couldn't figure another way around it.